### PR TITLE
lib: table: when using colSpan > 1 adjust the indexes when calculating data labels

### PR DESCRIPTION
**Pull request by KKoukiou**

### CODE REVIEW

File: pkg/lib/cockpit-components-table.jsx
Comment (allisonkarlitskaya) [inline pkg/lib/cockpit-components-table.jsx:234]: `curr.props?.colSpan || 1` ?

**Review by allisonkarlitskaya — DISMISSED** [review]
I'm not totally comfortable reviewing this (I don't even understand what the code does) but the logic seems reasonable.

I wonder if you could avoid the quadratic behaviour by using some form of map which can forward the cumulative count to the next iteration.  Something between reduce and map, I guess... Could also be accomplished by just open-coding it with a for loop...

So: go ahead and merge this if it's blocking you, but otherwise maybe wait for a more capable reviewer ;)

File: pkg/lib/cockpit-components-table.jsx
Comment (allisonkarlitskaya) [reply pkg/lib/cockpit-components-table.jsx:234]: Oh: and I suppose that colSpan won't ever be 0, right?  That wouldn't make sense... Because if it was 0, we'd end up rewriting it to 1.

Comment (cockpituous) [inline pkg/lib/cockpit-components-table.jsx:234]: This added line is not executed by any test.

Comment (KKoukiou) [reply pkg/lib/cockpit-components-table.jsx:234]: We dont specify 0 colspans. It's  >=1. colSpan 0 would rather be non present column.

**Comment by KKoukiou** [comment]
> I'm not totally comfortable reviewing this (I don't even understand what the code does) but the logic seems reasonable.
> 
> I wonder if you could avoid the quadratic behaviour by using some form of map which can forward the cumulative count to the next iteration. Something between reduce and map, I guess... Could also be accomplished by just open-coding it with a for loop...
> 
> So: go ahead and merge this if it's blocking you, but otherwise maybe wait for a more capable reviewer ;)

I can indeed remove the nested reduce. Will repost.

File: pkg/lib/cockpit-components-table.jsx
Comment (KKoukiou) [reply pkg/lib/cockpit-components-table.jsx]: it's covered by anaconda-webui tests.

File: pkg/lib/cockpit-components-table.jsx
Comment (allisonkarlitskaya) [inline pkg/lib/cockpit-components-table.jsx:235]: I find a map with side-effects to be a bit distasteful, but I'm not sure if it's common practice in JS or not...

Comment (allisonkarlitskaya) [reply pkg/lib/cockpit-components-table.jsx:235]: I went shopping in lodash for a primitive that would allow for something like this:

```ts
    const arr = [1, 2, 3];
    stateful_map(arr, item, state => [-(item + state), item], 0);
    > [-1, -3, -5]
```

ie: where each iteration allows returning a value but also allows forwarding a state value to the next iteration.

I couldn't find it, and I doubt you'd want to roll your own.

I defer to your judgment here.

Comment (KKoukiou) [reply pkg/lib/cockpit-components-table.jsx:235]: I would not like to complicate the logic. I dont think the current implementation is an antipractice and it's rather simple.

**Review by allisonkarlitskaya — APPROVED** [review]
Works for me :)

File: pkg/lib/cockpit-components-table.jsx
Comment (martinpitt) [reply pkg/lib/cockpit-components-table.jsx]: For the future: That's what `??` is for. `curr.props?.colSpan ?? 1` would  correctly return "0" if `colSpan == 0`.

**Pull request merged**
